OSDOCS#12034: HCP Kubevirt topology spread constraint (#90046)
openshift/openshift-docs: modules/hcp-topology-spread-constraint.adoc

// Module included in the following assemblies:
//
// * hosted_control_planes/hcp-manage/hcp-manage-virt.adoc

:_mod-docs-content-type: PROCEDURE
[id="hcp-topology-spread-constraint_{context}"]
= Spreading node pool VMs by using topologySpreadConstraint

> By default, KubeVirt virtual machines (VMs) created by a node pool are scheduled to any available node with the appropriate capacity to run the VM. A `topologySpreadConstraint` constraint spreads VMs of a node pool on multiple nodes. For example, if you are performing a live migration, you use `topologySpreadConstraint` with the descheduler so that VMs are continuously distributed on multiple nodes rather than a single node.

Reviewer: > For example, if you are performing a live migration, you use `topologySpreadConstraint` with the descheduler so that VMs are continuously distributed on multiple nodes rather than a single node.

The topologySpreadConstraint is always set by the hypershift controller when preparing the VMs for the node pool.
Nothing is really needed from the user.

The point is that for availability reason, the hypershift controller is setting a soft topologySpreadConstraint that under some circumstances (load peaks, unavailable nodes...) can lead to having two VMs of the same node pool running on the same physical machine with availability risks.
The descheduler with the `SoftTopologyAndDuplicates` profile can be used to get it periodically fixed.

.Prerequisites

* You installed the {descheduler-operator}. For more information, see "Installing the descheduler".

.Procedure

> * Modify the `KubeDescheduler` custom resource (CR) to use the `topologySpreadConstraint` constraint by running the following command:

Reviewer: to use the `SoftTopologyAndDuplicates` profile to ensure that the soft `topologySpreadConstraint` constraint is going to be satisfied again when possible.

// Module included in the following assemblies:
//
// * hosted_control_planes/hcp-manage/hcp-manage-virt.adoc

:_mod-docs-content-type: PROCEDURE
[id="hcp-topology-spread-constraint_{context}"]

> = Spreading node pool VMs by using topologySpreadConstraint

Author: Not sure if it's best to use back ticks in the heading.

> By default, KubeVirt virtual machines (VMs) created by a node pool are scheduled to any available nodes with the capacity to run the VM. By default, the `topologySpreadConstraint` constraint is set to schedule VMs on multiple nodes.

Reviewer 2: ```suggestion
By default, KubeVirt virtual machines (VMs) created by a node pool are scheduled on any available nodes that have the capacity to run the VMs. By default, the `topologySpreadConstraint` constraint is set to schedule VMs on multiple nodes.
```

> In some scenarios, VMs of a node pool might run on the same node which can cause an availability issue. To avoid distribution of VMs on a single node, use the descheduler to continuously honour the `topologySpreadConstraint` constraint to spread VMs on multiple nodes.

Reviewer 2: ```suggestion
In some scenarios, VMs for a node pool might run on the same node, which can cause availability issues. To avoid distribution of VMs on a single node, use the descheduler to continuously honour the `topologySpreadConstraint` constraint to spread VMs on multiple nodes.
```

> In some scenarios, VMs of a node pool might run on the same node which can cause an availability issue. To avoid distribution of VMs on a single node, use the `topologySpreadConstraint` constraint with the descheduler to spread VMs on multiple nodes when needed.

Reviewer: To avoid distribution of VMs on a single node, use the descheduler to keep honouring the `topologySpreadConstraint` constraint to spread VMs on multiple nodes when needed.

.Prerequisites

* You installed the {descheduler-operator}. For more information, see "Installing the descheduler".

.Procedure

> * Modify the `KubeDescheduler` custom resource (CR) to use the `SoftTopologyAndDuplicates` profile so that you maintain the `topologySpreadConstraint` constraint settings. Run the following command:

Reviewer 2: ```suggestion
* Open the `KubeDescheduler` CR  by entering the following commmand, and then modify the `KubeDescheduler` custom resource (CR) to use the `SoftTopologyAndDuplicates` profile so that you maintain the `topologySpreadConstraint` constraint settings
```

+
[source,terminal]
----
$ oc edit kubedescheduler cluster -n openshift-kube-descheduler-operator
----
+

> The `KubeDescheduler` CR named `cluster` runs in the `openshift-kube-descheduler-operator` namespace.

Reviewer 2: Consider moving "The `KubeDescheduler` CR named `cluster` runs in the `openshift-kube-descheduler-operator` namespace." above the command.

+
.Example `KubeDescheduler` configuration
[source,yaml]
----
apiVersion: operator.openshift.io/v1
kind: KubeDescheduler
metadata:
  name: cluster
  namespace: openshift-kube-descheduler-operator
spec:
  mode: Automatic
  managementState: Managed
  deschedulingIntervalSeconds: 60 // <1>
  profiles:
  - SoftTopologyAndDuplicates // <2>
  - EvictPodsWithPVC          // <3>
  - EvictPodsWithLocalStorage // <4>
  profileCustomizations:

>     devEnableEvictionsInBackground: true // <5>

Reviewer 2: ```suggestion
    devEnableEvictionsInBackground: true // <5>
# ...
```

If the YAML definition is not complete.

Author: Nice catch 😎

----
<1> Sets the number of seconds between the descheduler running cycles.

> <2> This profile evicts pods that follow the soft topology constraint, `whenUnsatisfiable: ScheduleAnyway`.

Reviewer 2: ```suggestion
<2> This profile evicts pods that follow the soft topology constraint: `whenUnsatisfiable: ScheduleAnyway`.
```

Author: Fixed
